_posts/2020-10-20-kafka-failure-strategy.adoc

```diff
@@ -0,0 +1,241 @@
+
+[source, bash]
+----
+jbang https://gist.github.com/cescoffier/23ca7b2bcc8c49cee3db29b3f2b59e4a/raw/0a1a8cd9a0cbed69d8025004cd5feab8c044d097/KafkaIgnoreFailure.java
+---- 
+
+If you ran the application and check the log, you will see two exceptions:
+[source]
+----
+ERROR [io.sma.rea.mes.provider] (vert.x-eventloop-thread-0) SRMSG00200: The method foo.KafkaFailFast$MovieProcessor#consume has thrown an exception: java.lang.IllegalArgumentException: I don't like movies with ' in their title: Schindler's List
+    at foo.KafkaFailFast$MovieProcessor.consume(KafkaFailFast.java:47)
+...
+ERROR [io.sma.rea.mes.provider] (vert.x-eventloop-thread-0) SRMSG00200: The method foo.KafkaIgnoreFailure$MovieProcessor#consume has thrown an exception: java.lang.IllegalArgumentException: I don't like movies with , in their title: The Good, the Bad and the Ugly
+    at foo.KafkaIgnoreFailure$MovieProcessor.consume(KafkaIgnoreFailure.java:51)    
+...
+WARN  [io.sma.rea.mes.kafka] (vert.x-eventloop-thread-0) SRMSG18204: A message sent to channel `movies` has been nacked, ignored failure is: I don't like movies with , in their title: The Good, the Bad and the Ugly.
+INFO  [Kafka-Ignore] (vert.x-eventloop-thread-0) Receiving movie The Lord of the Rings: The Fellowship of the Ring    
+----
+
+Look at the last line.
+As explained, it continues the processing with the next message.
+
+If you check the health of the application (using http://localhost:8080/health), everything is fine:
+
+[source, json]
+----
+{
+    "status": "UP",
+    "checks": [
+        {
+            "name": "SmallRye Reactive Messaging - liveness check",
+            "status": "UP",
+            "data": {
+                "movies": "[OK]",
+                "movies-out": "[OK]"
+            }
+        },
+        {
+            "name": "SmallRye Reactive Messaging - readiness check",
+            "status": "UP",
+            "data": {
+                "movies": "[OK]",
+                "movies-out": "[OK]"
+            }
+        }
+    ]
+}
+----
+
+== The "Dead-Letter Topic" strategy
+
+The https://en.wikipedia.org/wiki/Dead_letter_queue[dead-letter queue] is a well-known pattern to handle message processing failure. 
+Instead of failing fast or ignoring and continuing the processing, it stores the failing messages into a specific destination: a dead letter.
+An administrator, human or software, can review the failing messages and decide what can be done (retry, skip, etc.).
+Note that you can only apply this strategy if the ordering is not essential to the application.
+
+You can, later, review the failing messages:
+
+image:/assets/images/posts/kafka-failure-strategy/dead-letter-topic.png[Dead-letter topic]
+
+To enable this strategy, you need to add the following attribute to your configuration:
+[source, text]
+----
+mp.messaging.incoming.movies.failure-strategy=dead-letter-queue
+----
+
+By default, it writes to the `dead-letter-topic-$topic-name` topic.
```

Comment:
Is it worth mentioning whether this topic is created for them by choosing the DLQ strategy or if they need to create it themselves?

Reply:
Not related to the blog, but should the strategy be named DLQ if we're referring to it as DLT everywhere, or it doesn't matter?

Reply:
Yeah.. not really consistent. I realized that when writing. The pattern is named DLQ, but in Kafka we only have topics (that you can use as a queue). I need to improve it.

Reply:
Good point, I've added a note. Also, you may need to configure the replication factor.